app/models/patient.rb

```diff
@@ -742,23 +742,49 @@ def self.order_by_risk(asc: true)
   end
 
   # Check for potential duplicate records. Duplicate criteria is as follows:
-  # - matching first name, last name, sex, and DoB
+  # - matching values of first name and last name. Optionally sex and DoB can also match for more detailed messaging
+  # e.g. Jon Smith M null would be a duplicate of Jon Smith M 1/1/2000
   # OR
   # - matching state/local id
   def self.duplicate_data(first_name, last_name, sex, date_of_birth, user_defined_id_statelocal)
-    dup_info = where('first_name = ?', first_name)
-               .where('last_name = ?', last_name)
-               .where('sex = ?', sex)
-               .where('date_of_birth = ?', date_of_birth)
+    # Track which matches have occurred
+    duplicate_field_data = []
 
+    # check for a duplicate state/local id
     dup_statelocal_id = where('user_defined_id_statelocal = ?', user_defined_id_statelocal&.to_s&.strip)
+    duplicate_field_data << { count: dup_statelocal_id.size, fields: ['State/Local ID'] } if dup_statelocal_id.present?
+
+    # if first_name or last_name is null skip duplicate detection
+    return { is_duplicate: duplicate_field_data.any?, duplicate_field_data: duplicate_field_data } if first_name.nil? || last_name.nil?
+
+    fn_ln_sex_dob_matches = 0
+    fn_ln_sex_matches = 0
+    fn_ln_dob_matches = 0
+    fn_ln_matches = 0
+    potential_duplicates = where(first_name: first_name, last_name: last_name)
+
+    # Determine which type of duplicate exists
+    potential_duplicates.pluck(*%i[sex date_of_birth]).each do |p|
+      # If the sex isn't nil and doesn't match it is not a duplicate. Same for DoB.
+      if (sex.present? && !p[0].nil? && sex != p[0]) || (date_of_birth.present? && !p[1].nil? && date_of_birth != p[1].strftime('%F'))
```

Comment:
since `strftime('%F')` is being called multiple times, maybe we can write `dob = dob&.strftime('%F')` at the beginning of the loop. the `&` is for safe navigation when `dob` is `nil`